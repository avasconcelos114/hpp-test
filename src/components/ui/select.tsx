import * as React from 'react';
import Image from 'next/image';
import { cn } from '@/lib/utils';
import { Typography } from './typography';

export type SelectOption = {
  label: string;
  value: string;
};

interface SelectProps extends React.SelectHTMLAttributes<HTMLSelectElement> {
  label?: string;
  labelId?: string;
  helperText?: string;
  containerClassName?: string;
  options: SelectOption[];
  ariaLabel?: string;
}

const Select = React.forwardRef<HTMLSelectElement, SelectProps>(
  (
    {
      label,
      labelId,
      helperText,
      className,
      containerClassName,
      id,
      options,
      ariaLabel,
      ...props
    },
    ref,
  ) => {
    const generatedId = React.useId();
    const selectId = id || generatedId;
    const describedBy = helperText ? `${selectId}-helper-text` : undefined;
    const labelHtmlId = labelId || `${selectId}-label`;

    return (
      <div className={cn('flex w-full flex-col gap-1', containerClassName)}>
        {label && (
          <Typography
            id={labelHtmlId}
            htmlFor={selectId}
            size='md'
            weight='medium'
            tag='label'
            className='text-sm font-medium'
            ariaLabel={ariaLabel}
          >
            {label}
          </Typography>
        )}
        <div className='relative w-full'>
          <select
            id={selectId}
            ref={ref}
            aria-labelledby={label ? labelHtmlId : undefined}
            aria-describedby={describedBy}
            className={cn(
              // Hide native arrow, add right padding for icon
              'border-grays-line-gray focus:ring-ring/50 focus:border-ring h-[56px] w-full appearance-none rounded-md border px-3 py-2 pr-10 text-sm focus:ring-2 focus:outline-none disabled:cursor-not-allowed disabled:opacity-50',
              className,
            )}
            {...props}
          >
            {options.map((option) => (
              <option key={option.value} value={option.value}>
                {option.label}
              </option>
            ))}
          </select>
          <Image
            src='/icons/select_icon.svg'
            alt='Select dropdown icon'
            width={20}
            height={20}
            aria-hidden='true'
            className='pointer-events-none absolute top-1/2 right-3 -translate-y-1/2'
          />
        </div>
        {helperText && (
          <span id={describedBy} className={'text-muted-foreground text-xs'}>
            {helperText}
          </span>
        )}
      </div>
    );
  },
);
Select.displayName = 'Select';

export { Select };
