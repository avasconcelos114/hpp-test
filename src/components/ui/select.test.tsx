import React from 'react';
import { describe, it, expect, vi } from 'vitest';
import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import { Select } from './select';

describe('Select', () => {
  it('should render the select', () => {
    render(<Select options={[]} data-testid='select' />);
    expect(screen.getByTestId('select')).toMatchSnapshot();
  });

  it('renders options correctly', () => {
    const options = [
      { label: 'Option 1', value: '1' },
      { label: 'Option 2', value: '2' },
    ];
    render(<Select options={options} data-testid='select' />);
    const select = screen.getByTestId('select');
    expect(select.children.length).toBe(2);
    expect(screen.getByText('Option 1')).toBeDefined();
    expect(screen.getByText('Option 2')).toBeDefined();
  });

  it('renders label properly', () => {
    render(<Select label='My Label' options={[]} data-testid='select' />);
    expect(screen.getByText('My Label')).toBeDefined();
  });

  it('renders helperText and associates it with select', () => {
    render(<Select helperText='Help me' options={[]} data-testid='select' />);
    expect(screen.getByText('Help me')).toBeDefined();
    const select = screen.getByTestId('select');
    const help = screen.getByText('Help me');
    expect(select.getAttribute('aria-describedby')).toBe(help.id);
  });

  it('calls onChange when value changes', () => {
    const options = [
      { label: 'A', value: 'a' },
      { label: 'B', value: 'b' },
    ];
    const handleChange = vi.fn();
    render(
      <Select options={options} onChange={handleChange} data-testid='select' />,
    );
    const select = screen.getByTestId('select');
    fireEvent.change(select, { target: { value: 'b' } });
    expect(handleChange).toHaveBeenCalled();
    expect((select as HTMLSelectElement).value).toBe('b');
  });

  it('renders with custom class names', () => {
    render(
      <Select options={[]} className='custom-class' data-testid='select' />,
    );
    const select = screen.getByTestId('select');
    expect(select.className).toContain('custom-class');
  });
});
