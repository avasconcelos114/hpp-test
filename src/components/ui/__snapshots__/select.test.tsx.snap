// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > should render the select 1`] = `
<select
  class="border-grays-line-gray focus:ring-ring/50 focus:border-ring h-[56px] w-full appearance-none rounded-md border px-3 py-2 pr-10 text-sm focus:ring-2 focus:outline-none disabled:cursor-not-allowed disabled:opacity-50"
  data-testid="select"
  id="«r0»"
/>
`;
